File: day10.py
from collections import deque

def left(c): return c in "-J7"
def right(c): return c in "-FL"
def up(c): return c in "|JL"
def down(c): return c in "|F7"

def solve(part):
    grid = []
    while line := input():
        grid.append("."+line+".")
    
    m = len(grid[0])
    grid = ["."*m] + grid + ["."*m]
    n = len(grid)

    sr, sc = 0, 0
    for i in range(n):
        for j in range(m):
            if grid[i][j] == 'S':
                sr, sc = i, j

    if left(grid[sr][sc+1]) and right(grid[sr][sc-1]):
        grid[sr] = grid[sr][:sc] + '-' + grid[sr][sc+1:]
    if up(grid[sr+1][sc]) and down(grid[sr-1][sc]):
        grid[sr] = grid[sr][:sc] + '|' + grid[sr][sc+1:]
    if left(grid[sr][sc+1]) and down(grid[sr - 1][sc]):
        grid[sr] = grid[sr][:sc] + 'L' + grid[sr][sc+1:]
    if right(grid[sr][sc-1]) and down(grid[sr-1][sc]):
        grid[sr] = grid[sr][:sc] + 'J' + grid[sr][sc+1:]
    if right(grid[sr][sc-1]) and up(grid[sr+1][sc]):
        grid[sr] = grid[sr][:sc] + '7' + grid[sr][sc+1:]
    if left(grid[sr][sc+1]) and up(grid[sr+1][sc]):
        grid[sr] = grid[sr][:sc] + 'F' + grid[sr][sc+1:]

    isloop = [[False] * m for _ in range(n)]
    bfs = deque([(sr, sc)])
    steps = 0
    cpy = grid.copy()
    while bfs:
        steps += 1
        k = len(bfs)
        for _ in range(k):
            r, c = bfs.popleft()
            isloop[r][c] = True
            if left(cpy[r][c]) and cpy[r][c-1] != '.':
                bfs.append((r, c-1))
            if right(cpy[r][c]) and cpy[r][c+1] != '.':
                bfs.append((r, c+1))
            if up(cpy[r][c]) and cpy[r-1][c] != '.':
                bfs.append((r-1, c))
            if down(cpy[r][c]) and cpy[r+1][c] != '.':
                bfs.append((r+1, c))
            cpy[r] = cpy[r][:c] + '.' + cpy[r][c+1:]

    if part == 1:
        print(steps - 1)
    else:
        area = 0
        inloop = False
        lastpipe = '.'
        for i in range(n):
            for j in range(m):
                if isloop[i][j]:
                    if grid[i][j] in "|FL":
                        inloop = not inloop
                    elif grid[i][j] == '7' and lastpipe == 'F':
                        inloop = not inloop
                    elif grid[i][j] == 'J' and lastpipe == 'L':
                        inloop = not inloop

                    if grid[i][j] != '-':
                        lastpipe = grid[i][j]
                elif inloop:
                    area += 1

        print(area)

solve(1)
solve(2)

File: day10.cpp
#include <bits/stdc++.h>
using namespace std;
#define INPUT(var) for(string var; getline(cin, var), var!="";)
#define FOR(i, s, e) for (int i=s; i<e; ++i)

bool left(char c) { return string("-J7").find(c) != -1; }
bool right(char c) { return string("-FL").find(c) != -1; }
bool up(char c) { return string("|JL").find(c) != -1; }
bool down(char c) { return string("|F7").find(c) != -1; }

void solve(int part) {
	vector<string> grid;
	INPUT(line)
		grid.push_back("."+line+".");
	int m = size(grid[0]);
	grid.insert(begin(grid), string(m, '.'));
	grid.push_back(string(m, '.'));
	int n = size(grid);

	int sr, sc;
	FOR(i, 0, n)
		FOR(j, 0, m)
			if (grid[i][j] == 'S')
				sr = i, sc = j;

	if (left(grid[sr][sc+1]) && right(grid[sr][sc-1])) grid[sr][sc] = '-';
	if (up(grid[sr+1][sc]) && down(grid[sr-1][sc])) grid[sr][sc] = '|';
	if (left(grid[sr][sc+1]) && down(grid[sr-1][sc])) grid[sr][sc] = 'L';
	if (right(grid[sr][sc-1]) && down(grid[sr-1][sc])) grid[sr][sc] = 'J';
	if (right(grid[sr][sc-1]) && up(grid[sr+1][sc])) grid[sr][sc] = '7';
	if (left(grid[sr][sc+1]) && up(grid[sr+1][sc])) grid[sr][sc] = 'F';

	vector<vector<bool>> isloop(n, vector<bool>(m));
	queue<pair<int, int>> bfs;
	vector<string> cpy = grid;
	bfs.push({ sr, sc });
	int steps = 0;
	while (!bfs.empty()) {
		++steps;
		int k = bfs.size();
		while (k--) {
			auto [r, c] = bfs.front(); bfs.pop();
			isloop[r][c] = true;
			if (left(cpy[r][c]) && cpy[r][c-1] != '.') bfs.push({ r, c-1 });
			if (right(cpy[r][c]) && cpy[r][c+1] != '.') bfs.push({ r, c+1 });
			if (up(cpy[r][c]) && cpy[r-1][c] != '.') bfs.push({ r-1, c });
			if (down(cpy[r][c]) && cpy[r+1][c] != '.') bfs.push({ r+1, c });
			cpy[r][c] = '.';
		}
	}

	if (part == 1) {
		cout << steps-1 << '\n';
	}
	else {
		int area = 0;
		bool inloop = false;
		char lastpipe = '.';
		FOR(i, 0, n) {
			FOR(j, 0, m) {
				if (isloop[i][j]) {
					if (string("|FL").find(grid[i][j]) != -1) inloop = !inloop;
					else if (grid[i][j] == '7' && lastpipe == 'F') inloop = !inloop;
					else if (grid[i][j] == 'J' && lastpipe == 'L') inloop = !inloop;
					
					if (grid[i][j] != '-') lastpipe = grid[i][j];
				}
				else if (inloop) ++area;
			}
		}
		cout << area << '\n';
	}
}

int main() {
	solve(1);
	solve(2);
}